data_model/src/error.rs

```diff
@@ -0,0 +1,243 @@
+#[derive(Debug, Error, Decode, Encode, IntoSchema)]
+pub enum FindError {
+    /// Failed to find asset
+    #[error("Failed to find asset: `{0}`")]
+    Asset(AssetId),
+    /// Failed to find asset definition
+    #[error("Failed to find asset definition: `{0}`")]
+    AssetDefinition(AssetDefinitionId),
+    /// Failed to find account
+    #[error("Failed to find account: `{0}`")]
+    Account(<Account as Identifiable>::Id),
+    /// Failed to find domain
+    #[error("Failed to find domain: `{0}`")]
+    Domain(DomainId),
+    /// Failed to find metadata key
+    #[error("Failed to find metadata key")]
+    MetadataKey(Name),
+    /// Block with supplied parent hash not found. More description in a string.
+    #[error("Block with hash {0} not found.")]
+    Block(HashOf<VersionedCommittedBlock>),
+    /// Transaction with given hash not found.
+    #[error("Transaction not found")]
+    Transaction(HashOf<VersionedSignedTransaction>),
+    /// Value not found in context.
+    #[error("Value named {0} not found in context. ")]
+    Context(String),
+    /// Peer not found.
+    #[error("Peer {0} not found")]
+    Peer(PeerId),
+    /// Trigger not found.
+    #[error("Trigger not found.")]
+    Trigger(TriggerId),
+    /// Failed to find Role by id.
+    #[error("Failed to find role by id: `{0}`")]
+    Role(RoleId),
+    /// Failed to find [`PermissionToken`] by id.
+    #[error("Failed to find permission definition token by id: `{0}`")]
+    PermissionTokenDefinition(PermissionTokenId),
+    /// Failed to find [`Validator`](permission::Validator) by id.
+    #[error("Failed to find permission validator by id: `{0}`")]
+    Validator(permission::validator::Id),
+    /// Failed to find specified [`Parameter`] variant.
+    #[error("Failed to find specified parameter variant: `{0}`")]
+    Parameter(Parameter),
+}
+
+/// Generic structure used to represent a mismatch
+#[derive(Debug, Clone, PartialEq, Eq, Error, Decode, Encode, IntoSchema)]
+#[error("Expected {expected:?}, actual {actual:?}")]
+pub struct Mismatch<T> {
+    /// The value that is needed for normal execution
+    pub expected: T,
+    /// The value that caused the error
+    pub actual: T,
+}
+
+/// Type error
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[allow(variant_size_differences)] // False-positive
+pub enum TypeError {
+    /// Asset type assertion error
+    #[error("Asset Ids correspond to assets with different underlying types. {0}")]
+    AssetValueType(#[from] Mismatch<AssetValueType>),
+    /// Parameter type assertion error
+    #[error("Value passed to the parameter doesn't have the right type. {0}")]
+    ParameterValueType(#[from] Mismatch<Value>),
+    /// Asset Id mismatch
+    #[error("AssetDefinition Ids don't match. {0}")]
+    AssetDefinitionId(#[from] Box<Mismatch<<AssetDefinition as Identifiable>::Id>>),
+}
+
+/// Math error, which occurs during instruction execution
+#[derive(Debug, Clone, Error, Copy, PartialEq, Eq)]
+pub enum MathError {
```

Comment:
```suggestion
pub enum AssetMathError {
```

Reply:
We can do not only asset related operations inside iroha. So i think `MathError` is more accurate in this case.